src/main.rs

```diff
@@ -240,6 +271,129 @@ async fn main() -> Result<()> {
                 .collect::<Vec<_>>();
             println!("Groups found: {:?}", group_ids)
         }
+        CommandLineArgs::FileEncrypt {
+            filename: infile,
+            device_path,
+            users: user_ids,
+            groups: group_ids,
+            output: maybe_output,
+        } => {
+            let file = std::fs::read(infile.clone())?;
+            println!("Read in file {:?}", infile);
+            let sdk = initialize_sdk_from_file(&device_path).await?;
+            let output = match maybe_output {
+                // User specified an output path.
+                Some(desired) => {
+                    // User specified a directory for output with no filename
+                    if desired.is_dir() {
+                        // This is the file component of infile.
+                        let mut filename = match infile.file_name() {
+                            Some(name) => Result::Ok(name),
+                            None => Result::Err(InitAppErr(
+                                "Trying to write output to directory, but unable to infer file name"
+                                    .to_string(),
+                            )),
+                        }?
+                        .to_os_string();
+                        filename.push(".iron");
+                        let mut desired_filename = desired;
+                        desired_filename.push(filename);
+                        desired_filename
+                    } else {
+                        desired
+                    }
+                }
+                // User didn't specify an output path. Add ".iron" to the input path.
+                None => {
+                    let mut output_path = infile;
+                    let new_extension = match output_path.extension() {
+                        Some(extension) => {
+                            let mut ext_os = extension.to_os_string();
+                            ext_os.push(".iron");
+                            ext_os
+                        }
+                        None => std::ffi::OsString::from("iron"),
+                    };
+                    output_path.set_extension(new_extension);
+                    output_path
+                }
+            };
+            let mut users_or_groups = group_ids
+                .iter()
+                .map(|group| group.into())
+                .collect::<Vec<UserOrGroup>>();
+            let mut users = user_ids
+                .iter()
+                .map(|user| user.into())
+                .collect::<Vec<UserOrGroup>>();
+            users_or_groups.append(&mut users);
+            let grants = ExplicitGrant::new(true, &users_or_groups);
```

Comment:
Seems like from here to the bottom of this block should become an encrypt function. I really dislike that we've blown the main up to contain all the code. Small reusable functions seem best here.

Reply:
Moved into an `encrypt_file` function. I tried to keep the main small for these subcommands now